core/services/log/registrations.go

```diff
@@ -11,23 +11,23 @@ import (
 	"github.com/smartcontractkit/chainlink/core/store/models"
 )
 
-// How it works in general:
 // 1. Each listener being registered can specify a custom NumConfirmations - number of block confirmations required for any log being sent to it.
-// 2. Adding and removing listeners updates the highestNumConfirmations - a number tracking what's the current highest NumConfirmations globally
 //
-// 3. All received logs are kept in an array and deleted ONLY after they are outside the confirmation range for all subscribers
+// 2. All received logs are kept in an array and deleted ONLY after they are outside the confirmation range for all subscribers
 // (when given log height is lower than (latest height - max(highestNumConfirmations, ETH_FINALITY_DEPTH)) ) -> see: pool.go
 //
+// 3. Information about already consumed logs is fetched from the database and used as a filter
+//
 // 4. The logs are attempted to be sent after every new head arrival:
-// 		Each stored log is then checked against every matched listener and is sent unless:
+// 		Each stored log is checked against every matched listener and is sent unless:
 //    A) is too young for that listener
-//    B) the corresponding block height is known to be already processed for that listener
-// In the normal case, each log will be only processed once, and then its corresponding head will be remembered, so it's not double-sent
+//    B) matches a log already consumed (via the database information from log_broadcasts table)
 //
-// After processing the whole batch of logs considered for sending, the per-listener metadata is updated in applyListenerInfoUpdates.
-// If a re-org happens, the stored lowestAllowedBlockNumber (per-listener) is re-set,
-// so the logs from that chain are then considered unprocessed, and will be sent again.
+// A log might be sent multiple times, if a consumer processes logs asynchronously (e.g. via a queue or a Mailbox), in which case the log
+// may not be marked as consumed before the next sending operation. That's why customers must still check the state via WasAlreadyConsumed
+// before processing the log.
```

Comment:
Ideally the consumer would be operating sequentially - is that not the case?

Reply:
no, most/all consumers just put the log in a mailbox. Which makes sense i think, otherwise LB processing would be much more slowed down by waiting for each subscriber / would need more parallelisation